Update typedoc to fix a security vulnerability

devalevenkatesh opened the pull request:
**Issue #:**
Fix a security vulnerability with `marked` library. We depend on `typedoc` for our doc generation which depends on `marked`. `typedoc's` latest version fixes the security vulnerability with `marked` hence we need to bump our `typedoc` from `0.19.2` to `0.20.25`.

**Description of changes:**
- Updated `typedoc` to fix a security vulnerability associated with the `marked` library.
- Followed the migration guide under `0.20.0` release from [here](https://github.com/TypeStrong/typedoc/releases?after=v0.20.1) and the typedoc [options](https://typedoc.org/guides/options/) guide.

**Testing**

1. Have you successfully run `npm run build:release` locally? Yes.
2. How did you test these changes? 
- By opening up the generated docs using `open index.html` after `npm run build` and `npm run doc`.
- There are some decorative changes I found since we bumped `typedoc` from `0.19.2` to `0.20.25`. But, all the documentation content looked similar to what was generated using the `typedoc` `0.19.2` version.
3. Can these changes be tested using the demo application? If yes, which demo application can be used to test it? No. This is a documentation package update.
4. Do you add, modify, or delete public API definitions? If yes, has that been reviewed and approved? NA.
5. Do you change the wire protocol, e.g. the request method? If yes, has that been reviewed and approved? NA.


By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.

ltrung commented:
Do we want to update all the doc filenames?

ltrung commented:
since we could specify typedoc config inside tsconfig.js, can we get rid of typedoc.js (and maybe the one inside guides folder)?

devalevenkatesh commented:
> since we could specify typedoc config inside tsconfig.js, can we get rid of typedoc.js (and maybe the one inside guides folder)?

Yup, so I have removed `typedoc.json` and added those options under `guides/tsconfig.json` under `typedocOptions` key. Also, we need two separate `tsconfig.json`s since our `typedoc` uses different set of keys than the build process keys.

devalevenkatesh commented:
> Do we want to update all the doc filenames?

According to new typedoc version they removed the support for "mode" option and introduced "entryPoints" option. If we have multiple entry points (our case), we get the module name appended to the filenames. I have added both "src/index.ts" and "guides/index.ts" as entry points but with custom module names. The exports from "src/index.ts" will be shown under "amazon-chime-sdk-js" module and similarly exports from "guides/index.ts" will show under "Guides".

ltrung commented on script/generate-docs:
```diff
@@ -19,4 +19,4 @@ STDERR.puts "Generating docs against #{remote}/#{branch}."
 # both the source and docs. The build itself just uses the root tsconfig,
 # which just hits src.
-`node node_modules/typedoc/bin/typedoc --tsconfig guides/tsconfig.json --options typedoc.json --gitRemote #{remote} --gitRevision #{branch}`
+`node node_modules/typedoc/bin/typedoc --tsconfig guides/tsconfig.json --gitRemote #{remote} --gitRevision #{branch}`
```
For the library src folder does it recognize it automatically since you only specific the one from the guides folder here?

devalevenkatesh replied:
The library src folder is picked from the typedoc options in the guides/tsconfig.json, we have specified two entry points "src/index.ts" and "guides/index.ts". Thats how it recognizes the two.

ltrung commented on script/build-guides.js:
```diff
@@ -46,9 +47,17 @@ walk('guides')
+ * @packageDocumentation
+ * @module Guides
+ */`;
+fs.writeFileSync('./guides/index.ts', `${customModuleNameAnnotation}\n${namespaceExports} from './docs';`, 'utf8');
```
What is the difference between index.ts and doc.ts?

devalevenkatesh replied:
To give the "guides/docs.ts" a different custom name like "Guides", I had to export the namespaces that are generated at runtime in the "docs.ts". Once exported I had to re-export them from "index.ts" where I could give a custom module name to that "index.ts". Giving a custom name at the start of the "docs.ts" did not helped as that file complains of not being a module. And I tried various variations but still the name in generated docs remained "guides/docs.ts" and not what I provided. So as Richard suggested to have index.ts and export everything from docs.ts, that helped.
